### DISCUSSION TITLE
How to enable cpu targets on a cluster without root access?

### DISCUSSION THREAD

**Discussion by csccva** [intel/llvm#12360] [General]
Hello,

I managed to compile intel llvm on a cluster with nvidia gpus. Everything works for nvidia targets, but I am not able to use cpu targets. 
From [these instructions](https://intel.github.io/llvm-docs/GetStartedGuide.html) I figure that the opencl runtime for cpus is needed, but I can figure out how to integrate it my build as a normal user.

 EDIT:

I tried to use as target `native_cpu`, but this option does not find any device. I get this error:

```
terminate called after throwing an instance of 'sycl::_V1::runtime_error'
  what():  No device of requested type available. -1 (PI_ERROR_DEVICE_NOT_FOUND)
```

**Comment by AlexeySachkov** [comment]
Hi @csccva,

> From [these instructions](https://intel.github.io/llvm-docs/GetStartedGuide.html) I figure that the opencl runtime for cpus is needed, but I can figure out how to integrate it my build as a normal user.

Our OpenCL CPU Runtime supports loading through [Khronos OpenCL ICD Loader](https://github.com/KhronosGroup/OpenCL-ICD-Loader) (and that's in fact is the main way of how it should be used). The flow here is that SYCL RT through some so-called plugins or adapters libraries is linked to `libOpenCL.so` (ICD Loader) which is in turn responsible for finding actual OpenCL implementations.

Note that the runtime itself can be placed in any directory as long as OpenCL ICD Loader can discover it.

As you can see in [Registering ICDs](https://github.com/KhronosGroup/OpenCL-ICD-Loader?tab=readme-ov-file#registering-icds), default locations where the loader is looking for instructions where to find runtimes require access to `/opt` area. However, you can scroll down to [Table of debug environment variables](https://github.com/KhronosGroup/OpenCL-ICD-Loader?tab=readme-ov-file#table-of-debug-environment-variables) and there you can find more controls over where the loader looks for instructions to find OpenCL implementations.

Those variables are called debug ones, but in fact they are pretty stable and we have been using them in our own testing environments for quite a while. So, this is probably a not 100% production-quality solution, but it should be stable and reliable enough.

There is an alternative, but way more hacky way. If I'm not mistaken, `libintelocl.so` library (entry point for OpenCL CPU RT) implements the same API/interface as `libOpenCL.so`. Therefore, what you can do is to rename `libintelocl.so` into `libOpenCL.so` (or create a symlink) and make sure that it is the first one to be picked up (i.e. configure `LD_LIBRARY_PATH` or use `LD_PRELOAD`). This way, SYCL RT will directly use OpenCL CPU RT by-passing ICD Loader. There are two downside of this approach:
1. we do not regularly test this flow, there may be some issues/problems which are unknown to us;
2. OpenCL CPU RT does not work as the loader, meaning that this way you won't be able to use _any_ other OpenCL CPU RT which may be also installed on the system;

Hope this helps

**Comment by csccva** [comment]
Thank you for your reply.

I started to do as you instructed, but during the process I found out that I can actually install everything I needed. 
```
//Download oneapi offline 
wget https://registrationcenter-download.intel.com/akdlm/IRC_NAS/163da6e4-56eb-4948-aba3-debcec61c064/l_BaseKit_p_2024.0.1.46_offline.sh 
//Install in silent mode in a non-standard folder
chmod +x l_BaseKit_p_2024.0.1.46_offline.sh
./l_BaseKit_p_2024.0.1.46_offline.sh  -a -s --eula accept --download-cache <temp_path>/tttt/ --install-dir <path>/intel/oneapi 
//Download plugin for nvidia
curl -LOJ "https://developer.codeplay.com/api/v1/products/download?product=oneapi&variant=nvidia&version=2024.0.1&filters[]=12.0&filters[]=linux"
 //Install in the oneapi folder
./oneapi-for-nvidia-gpus-2024.0.1-cuda-12.0-linux.sh -y --extract-folder <temp_path>/tttt/ --install-dir <path>/intel/oneapi
//For usage first set the environment right
. /scratch/project_2008874/cristian/intel/oneapi/setvars.sh --include-intel-llvm
// Compile the code with nvidia and cpu targets
module load cuda
 clang++ -std=c++17 -O3 -fsycl -fsycl-targets=nvptx64-nvidia-cuda,spir64_x86_64 -Xsycl-target-backend=nvptx64-nvidia-cuda --cuda-gpu-arch=sm_80  <sycl_code>.cpp

//Run the code 
```

**Comment by csccva** [comment]
Hello,

Sorry for the delay in answering I thought I replied already here. I managed to do what I wanted by installing the Intel base kit 2024 and then I downloaded from codeplay the nvidia/amd plug ins and apply the patch. This way I was able to get mky sycl codes running on CPUs and also on nvidia or amd GPUs. 

Cristian
